README.md
# Caesar and Vigenere ciphers

Hello everyone! This is the repository of my package on JavaScript "Caesar and Vigenere ciphers"

## Table of contents
  * [Motivation](#motivation)
  * [Build status](#build-status)
  * [Code style](#code-style)
  * [Dependencies](#dependencies)
  * [Features](#features)
  * [Installation](#installation)
  * [Importing](#importing)
  * [Fast usage](#fast-usage)
  * [API](#api)
  * [Code Example](#code-example)
  * [Tests](#tests)
  * [Contributing](#contributing)
  * [License](#license)

## Motivation

I've always been interested in the topic of data encryption, but I didn't delve deeply into it. So I decided to do it on. At first there was an idea to do a [*CLI*](https://en.wikipedia.org/wiki/Command-line_interface) or an *[Electron](https://www.electronjs.org/) application*. But later it turned into a package for [*npm*](https://www.npmjs.com/).

## Build status

Here you can see build status of [continuous integration](https://en.wikipedia.org/wiki/Continuous_integration)/[continuous deployment](https://en.wikipedia.org/wiki/Continuous_deployment):

![Node.js Package](https://github.com/mezgoodle/Caesar-and-Vigenere-ciphers/workflows/Node.js%20Package/badge.svg)
[![Build Status](https://travis-ci.com/mezgoodle/Caesar-and-Vigenere-ciphers.svg?branch=master)](https://travis-ci.com/mezgoodle/Caesar-and-Vigenere-ciphers)
[![pipeline status](https://gitlab.com/Zavalniuk/Caesar-and-Vigenere-ciphers/badges/master/pipeline.svg)](https://gitlab.com/Zavalniuk/Caesar-and-Vigenere-ciphers/-/commits/master)
[![Build Status](https://dev.azure.com/mezgoodle/mezgoodle/_apis/build/status/mezgoodle.Caesar-and-Vigenere-ciphers?branchName=master)](https://dev.azure.com/mezgoodle/mezgoodle/_build/latest?definitionId=2&branchName=master)
![npm](https://img.shields.io/npm/dt/@mezgoodle/caesar-and-vigenere-ciphers)
![node-current](https://img.shields.io/node/v/jest)

## Code style

I'm using [Codacy](https://www.codacy.com/) for automate my code quality.

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/c4a0fd48cea64b4aa06936c12d853856)](https://www.codacy.com/manual/mezgoodle/Caesar-and-Vigenere-ciphers?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=mezgoodle/Caesar-and-Vigenere-ciphers&amp;utm_campaign=Badge_Grade)

## Dependencies

![David](https://img.shields.io/david/dev/mezgoodle/Caesar-and-Vigenere-ciphers)

I'm using only [jest](https://jestjs.io/) for testing, but that's enough for unit-testing. You may not install it.

>You can see all dependencies from `package-lock.json` [here](https://github.com/mezgoodle/Caesar-and-Vigenere-ciphers/network/dependencies).

## Features

With my package you can encrypt / decrypt text with Caesar and Wiegener  algorithms with *latin*, *cyrillic* or *greek* alphabet based on **UTF-16**.

> If you see incorrect work with alphabet, please write an issue.

## Installation

First install [Node.js](https://nodejs.org/uk/). Then:

```bash
npm i @mezgoodle/caesar-and-vigenere-ciphers --save
```

## Importing

```js
// Using Node.js `require()`
const { caesarEncrypt, caesarDecrypt, vigenereEncrypt, vigenereDecrypt } = require("@mezgoodle/caesar-and-vigenere-ciphers");
```

## Fast usage

```js
// Latin

console.log(caesarEncrypt("pellentesque", 12, "lat"));
// expected output: bqxxqzfqecgq
console.log(caesarDecrypt("pellentesque", -12, "lat"));
// expected output: bqxxqzfqecgq
console.log(vigenereEncrypt("unopinionated", "express", "lat"));
// expected output: ykdgmfaskpkiv
console.log(vigenereDecrypt("ykdgmfaskpkiv", "express", "lat"));
// expected output: unopinionated

// Cyrillic

console.log(caesarEncrypt("привет", 2, "cyr"));
// expected output: сткдзф
console.log(caesarDecrypt("сткдзф", 2, "cyr"));
// expected output: привет
console.log(vigenereEncrypt("Карл у Клары украл кораллы", "кларнет", "cyr"));
// expected output: Флры а Пэкыы гчхтх хоанрэе
console.log(vigenereDecrypt("Флры а Пэкыы гчхтх хоанрэе", "кларнет", "cyr"));
// expected output: Карл у Клары украл кораллы

// Greek

console.log(caesarEncrypt("αβ", 1, "gre"));
// expected output: βγ
console.log(caesarDecrypt("βγ", 1, "gre"));
// expected output: αβ
console.log(vigenereEncrypt("ΓειΑ", "βι", "gre"));
// expected output: ΔνκΙ
console.log(vigenereDecrypt("ΔνκΙ", "βι", "gre"));
// expected output: ΓειΑ
```

## API

### caesarEncrypt( value, amount, type )

Name    | Type     | Argument     | Default | Description
--------|----------|--------------|---------|------------
value     | `string` | `<required>` | `null`  | the message to encrypt
amount | `number` | `<required>` | `null`  | the key to encrypt the message with
type | `string` | `<required>` | `null`  | the type of language: latin or cyrillic

### caesarDecrypt( value, amount, type )

Name    | Type     | Argument     | Default | Description
--------|----------|--------------|---------|------------
value     | `string` | `<required>` | `null`  | the message to decrypt
amount | `number` | `<required>` | `null`  | the key to decrypt the message with
type | `string` | `<required>` | `null`  | the type of language: latin or cyrillic

### vigenereEncrypt( text, key, type )

Name    | Type     | Argument     | Default | Description
--------|----------|--------------|---------|------------
text     | `string` | `<required>` | `null`  | the message to encrypt
key | `string` | `<required>` | `null`  | the key to encrypt the message with
type | `string` | `<required>` | `null`  | the type of language: latin or cyrillic

### vigenereDecrypt( text, key, type )

Name    | Type     | Argument     | Default | Description
--------|----------|--------------|---------|------------
text     | `string` | `<required>` | `null`  | the message to decrypt
key | `string` | `<required>` | `null`  | the key to decrypt the message with
type | `string` | `<required>` | `null`  | the type of language: latin or cyrillic


## Code Example

Here you can see small example of Vigenere worker

```js
const worker = (str, key, type, lang) => {
  if (typeof(key) !== 'string' || typeof(str) !== 'string')
    throw Error('Text and Key must be string');
  if (!Object.prototype.hasOwnProperty.call(typeDefine, lang))
    throw Error('Type must be "lat" or "cyr"');
  if (str === null || key === null) throw Error('Message and key should be not empty');
  key = keepLetters(key);
  let result = '',
    keyIndex = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charAt(i);
    if (isLetter(char)) {
      const key_el = key.charAt(keyIndex++ % key.length);
      const temp = workerChar(char, key_el, type, lang);
      result += temp;
    } else {
      result += char;
    }
  }
  return result;
};
```

## Tests

Unit-testing with **assert** and **jest**:

>Early here were screenshots

Number of my tests is more than 50, so I can't just do screenshot :smile:. I give you the [link](https://travis-ci.com/github/mezgoodle/Caesar-and-Vigenere-ciphers) to [Travis CI](https://travis-ci.org/), where you can see all my tests.

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change. Also look at the [CONTRIBUTING.md](https://github.com/mezgoodle/Caesar-and-Vigenere-ciphers/blob/master/CONTRIBUTING.md).

## License

![GitHub](https://img.shields.io/github/license/mezgoodle/Caesar-and-Vigenere-ciphers)

MIT © [mezgoodle](https://github.com/mezgoodle)


CHANGELOG.md
2.1.0 (02.06.2020)

- Added functionality that allows to encrypt and decrypt text in Greek using Caesar and Vigener ciphers.

2.0.3 (23.05.2020)

- Fix bugs.
- Fix lint errors.

2.0.0 (23.05.2020)

- Added functionality that allows to encrypt and decrypt text in Cyrillic using Caesar and Vigener ciphers.

1.1.6 (19.05.2020)

- Added one more conditional in Vigenere worker for checking empty strings.
- Added new tests for Vigenere cipher.

1.1.3 (09.05.2020)

- Improved some functions and delete duplicated code, so package is less.

1.1.0 (07.05.2020)

- Create Vigenere encrypt/decrypt functionality.

1.0.10 (06.05.2020)

- Add CD to GitHub Packages via GitHub Actions

1.0.0 (05.05.2020)

- Create Caesar encrypt/decrypt functionality.

test.js
const { caesarEncrypt } = require("./ciphers");
const assert = require("assert").strict;

// Test
const tests = [
    [
        ["pellentesque", 12], "bqxxqzfqecgq", "testSimple"
    ],
    [
        ["pellentesque", 0], "pellentesque", "testZeroAmount"
    ],
    [
        ["bqxxqzfqecgq", -12], "pellentesque", "testSimpleReverse"
    ],
    [
        ["adadA", 13123], "twtwT", "testUpperLowerCase"
    ],
    [
        ["ad_dA", 1], "be_eB", "testStringWithSymbols"
    ],
    [
        ["13123131", 5], "13123131", "testIntegers"
    ],
    [
        ["!@&*^", -134353], "!@&*^", "testSymbols"
    ],
    [
        ["!@$^@a", -1970], "!@$^@g", "testSymbolsWithString"
    ],
    [
        ["dada", -19.70], "jgjg", "testFloatAmount"
    ],
];

// Output
const results = [];
for (const test of tests) {
    const [par, expected, name] = test;
    const result = caesarEncrypt(par[0], par[1]);
    try {
        assert.strictEqual(result, expected, `Error in test "${name}"`);
    } catch (err) {
        const { message, operator } = err;
        results.push({ message, par, expected, result, operator });
    }
}
console.table(results);

tests/data.js
// Tests
const tests = [
    [
        { value: "pellentesque", amount: 12 }, "bqxxqzfqecgq", "testSimple"
    ],
    [
        { value: "pellentesque", amount: 0 }, "pellentesque", "testZeroAmount"
    ],
    [
        { value: "bqxxqzfqecgq", amount: -12 }, "pellentesque", "testSimpleReverse"
    ],
    [
        { value: "adadA", amount: 13123 }, "twtwT", "testUpperLowerCase"
    ],
    [
        { value: "ad_dA", amount: 1 }, "be_eB", "testStringWithSymbols"
    ],
    [
        { value: "13123131", amount: 5 }, "13123131", "testIntegers"
    ],
    [
        { value: "!@&*^", amount: -134353 }, "!@&*^", "testSymbols"
    ],
    [
        { value: "!@$^@a", amount: -1970 }, "!@$^@g", "testSymbolsWithString"
    ],
    [
        { value: "dada", amount: -19.70 }, "jgjg", "testFloatAmount"
    ],
];

module.exports = tests;

tests/tests.js
'use strict';

const {
  caesarEncrypt,
  caesarDecrypt,
  vigenereEncrypt,
  vigenereDecrypt
} = require('../index');

const data = require('./data.json');
const assert = require('assert').strict;

const testingWorker = (data, fn) => {
  const results = [];
  for (const test of data) {
    const [params, expected, name] = test;
    const result = fn(params.value, params.key, params.type);
    try {
      assert.strictEqual(result, expected, `Error in test "${name}"`);
    } catch (err) {
      const { message, operator } = err;
      results.push({ message, params, expected, result, operator });
    }
  }
  return results;
};

// Output
let results = [];

const config = [{
  name: 'Caesar Encrypt',
  data: data.caesareEncrypt,
  fn: caesarEncrypt
},
{
  name: 'Caesar Decrypt',
  data: data.caesareDecrypt,
  fn: caesarDecrypt
},
{
  name: 'Vigenere Encrypt',
  data: data.vigenereEncrypt,
  fn: vigenereEncrypt
},
{
  name: 'Vigenere Decrypt',
  data: data.vigenereDecrypt,
  fn: vigenereDecrypt
}
];

// Do all tests from config
for (const element of config) {
  console.log(element.name);
  results = testingWorker(element.data, element.fn);
  console.table(results);
  results = [];
}


ciphers.js
const caesarEncrypt = (str = null, amount = null) => {

    if (str === null) {
        throw Error("Message should be not empty");
    }

    if (!Number.isInteger(amount) || amount === null) {
        throw Error("Amount should be integer and not empty");
    }

    // Wrap the amount
    if (amount < 0) {
        return caesarEncrypt(str, amount + 26);
    }

    if (typeof(str) !== "string") {
        throw Error("String or number expected");
    }

    // Make an output letiable
    let output = "";

    // Go through each character
    for (let i = 0; i < str.length; i++) {

        // Get the character we'll be appending
        let c = str[parseInt(i)];

        // If it's a letter...
        if (c.match(/[a-z]/i)) {

            // Get its code
            let code = str.charCodeAt(i);

            // Uppercase letters
            if ((code >= 65) && (code <= 90)) {
                c = String.fromCharCode(((code - 65 + amount) % 26) + 65);
            }

            // Lowercase letters
            else if ((code >= 97) && (code <= 122)) {
                c = String.fromCharCode(((code - 97 + amount) % 26) + 97);
            }

        }

        // Append
        output += c;

    }

    // All done!
    return output;
};

const caesarDecrypt = (text, shift) => {
    let result = "";
    shift = (26 - shift) % 26;
    result = caesarEncrypt(text, shift);
    return result;
};
module.exports = { caesarEncrypt, caesarDecrypt };

index.js
'use strict';

const typeDefine = {
  'cyr': { 'AlphaNum': 32, 'StartCodeU': 1040,
    'FinishCodeU': 1071, 'StartCodeL': 1072, 'FinishCodeL': 1105 },
  'lat': { 'AlphaNum': 26, 'StartCodeU': 65,
    'FinishCodeU': 90, 'StartCodeL': 97, 'FinishCodeL': 122 },
  'gre': { 'AlphaNum': 25, 'StartCodeU': 913,
    'FinishCodeU': 937, 'StartCodeL': 945, 'FinishCodeL': 969 },
};

const keepLetters = s => (s.replace(/[^a-zA-Zа-яА-Яα-ωΑ-Ω]+/g, ''));
const isLetter = c => (c.match(/[a-zA-Zа-яА-Яα-ωΑ-Ω]+/));
const isUpperCase = c => (c.match(/[A-ZА-ЯΑ-Ω]+/));

const caesarEncrypt = (text = null, amount = null, type = null) => {
  if (typeof amount !== 'number') {
    throw Error('Amount should be integer and not empty');
  }
  if (text === null) throw Error('Message should be not empty');
  if (typeof text !== 'string') throw Error('Text must be String');
  // Type define
  if (!Object.prototype.hasOwnProperty.call(typeDefine, type)) {
    throw Error('Type must be "lat" or "cyr"');
  }
  const {
    AlphaNum,
    StartCodeU,
    FinishCodeU,
    StartCodeL,
    FinishCodeL
  } = typeDefine[type];
  // Wrap the amount
  if (amount < 0) return caesarEncrypt(text, amount + AlphaNum, type);

  let output = '';
  for (let i = 0; i < text.length; i++) {
    let char = text[i];
    if (isLetter(char)) {
      // Get its code
      const code = text.charCodeAt(i);
      // lowercase letters
      if ((code >= StartCodeL) && (code <= FinishCodeL))
        char = String.fromCharCode(
          ((code - StartCodeL + amount) % AlphaNum) + StartCodeL);
      // Uppercase letters
      else if ((code >= StartCodeU) && (code <= FinishCodeU)) {
        char = String.fromCharCode(
          ((code - StartCodeU + amount) % AlphaNum) + StartCodeU);
      }
    }
    output += char;
  }
  return output;
};

const caesarDecrypt = (text, shift, type) => {
  if (!Object.prototype.hasOwnProperty.call(typeDefine, type)) {
    throw Error('Type must be "lat" or "cyr"');
  }
  const { AlphaNum } = typeDefine[type];
  let result = '';
  shift = (AlphaNum - shift) % AlphaNum;
  result = caesarEncrypt(text, shift, type);
  return result;
};

// Worker for single char in Vigenere algorithm
const workerChar = (char, key, type = 'e', lang) => {
  if (!isLetter(char) || !isLetter(key)) return char;
  const uppercase = isUpperCase(char);
  char = char.toLowerCase().charCodeAt(0);
  key = key.toLowerCase().charCodeAt(0);
  const { AlphaNum, StartCodeU, StartCodeL } = typeDefine[lang];
  const a = StartCodeL;
  const A = StartCodeU;
  if (type === 'd') {
    let temp = a + ((char - a) - (key - a));
    if (temp < a) temp += AlphaNum;
    temp = String.fromCharCode(temp);
    return uppercase ? temp.toUpperCase() : temp;
  } else return String.fromCharCode(
    (uppercase ? A : a) + (((char - a) + (key - a)) % AlphaNum));
};

// Worker for Vigenere algorithm
const worker = (str, key, type, lang) => {
  if (typeof key !== 'string' || typeof str !== 'string') {
    throw Error('Text and Key must be string');
  }
  if (!Object.prototype.hasOwnProperty.call(typeDefine, lang)) {
    throw Error('Type must be "lat" or "cyr"');
  }
  if (str === null || key === null) {
    throw Error('Message and key should be not empty');
  }
  key = keepLetters(key);
  let result = '';
  let keyIndex = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charAt(i);
    if (isLetter(char)) {
      const keyEl = key.charAt(keyIndex++ % key.length);
      const temp = workerChar(char, keyEl, type, lang);
      result += temp;
    } else {
      result += char;
    }
  }
  return result;
};

const vigenereEncrypt = (text = null, key = null, lang = null) =>
  (worker(text, key, 'e', lang));

const vigenereDecrypt = (cipher = null, key = null, lang = null) =>
  (worker(cipher, key, 'd', lang));

module.exports = {
  caesarEncrypt,
  caesarDecrypt,
  vigenereEncrypt,
  vigenereDecrypt
};
